common/env/env.go

```diff
@@ -0,0 +1,393 @@
+	if lib == nil {
+		return fn, true
+	}
+	if len(lib.IncludeFunctions) != 0 {
+		for _, include := range lib.IncludeFunctions {
+			if include.Name != fn.Name() {
+				continue
+			}
+			if len(include.Overloads) == 0 {
+				return fn, true
+			}
+			overloadIDs := make([]string, len(include.Overloads))
+			for i, o := range include.Overloads {
+				overloadIDs[i] = o.ID
+			}
+			return fn.Subset(decls.IncludeOverloads(overloadIDs...)), true
+		}
+		return nil, false
+	}
+	if len(lib.ExcludeFunctions) != 0 {
+		for _, exclude := range lib.ExcludeFunctions {
+			if exclude.Name != fn.Name() {
+				continue
+			}
+			if len(exclude.Overloads) == 0 {
+				return nil, false
+			}
+			overloadIDs := make([]string, len(exclude.Overloads))
+			for i, o := range exclude.Overloads {
+				overloadIDs[i] = o.ID
+			}
+			return fn.Subset(decls.ExcludeOverloads(overloadIDs...)), true
+		}
+		return fn, true
+	}
+	return fn, true
+}
+
+// SubsetMacro indicates whether the macro function should be included in the library subset.
+func (lib *LibrarySubset) SubsetMacro(macroFunction string) bool {
+	// When lib is null, it should indicate that all values are included in the subset.
+	if lib == nil {
+		return true
+	}
+	if lib.DisableMacros {
+		return false
+	}
+	if len(lib.IncludeMacros) != 0 {
+		for _, name := range lib.IncludeMacros {
+			if name == macroFunction {
+				return true
+			}
+		}
+		return false
+	}
+	if len(lib.ExcludeMacros) != 0 {
+		for _, name := range lib.ExcludeMacros {
+			if name == macroFunction {
+				return false
+			}
+		}
+		return true
+	}
+	return true
+}
+
+// TypeDesc represents the serializable format of a CEL *types.Type value.
+type TypeDesc struct {
+	TypeName    string      `yaml:"type_name"`
+	Params      []*TypeDesc `yaml:"params"`
+	IsTypeParam bool        `yaml:"is_type_param"`
+}
+
+// AsCELType converts the serializable object to a *types.Type value.
+func (td *TypeDesc) AsCELType(tp types.Provider) (*types.Type, error) {
+	if td == nil {
+		return nil, errors.New("nil TypeDesc cannot be converted to a Type instance")
+	}
+	if td.TypeName == "" {
+		return nil, errors.New("invalid type description, declare a type name")
+	}
+	var err error
+	switch td.TypeName {
+	case "dyn":
+		return types.DynType, nil
+	case "map":
+		if len(td.Params) == 2 {
+			kt, err := td.Params[0].AsCELType(tp)
+			if err != nil {
+				return nil, err
+			}
+			vt, err := td.Params[1].AsCELType(tp)
+			if err != nil {
+				return nil, err
+			}
+			return types.NewMapType(kt, vt), nil
+		}
+		return nil, fmt.Errorf("map type has unexpected param count: %d", len(td.Params))
+	case "list":
+		if len(td.Params) == 1 {
+			et, err := td.Params[0].AsCELType(tp)
+			if err != nil {
+				return nil, err
+			}
+			return types.NewListType(et), nil
+		}
+		return nil, fmt.Errorf("list type has unexpected param count: %d", len(td.Params))
+	case "optional", "optional_type":
```

Comment:
`optional_type` is the actual type name for optionals right? Should we just remove the case for `optional`, or is this to maintain backwards compatibility?

Reply:
I'll just remove the back-compat, I don't think it's necessary